server/src/main/java/org/elasticsearch/search/vectors/RescoreKnnVectorQuery.java

```diff
@@ -0,0 +1,160 @@
+
+import org.apache.lucene.index.VectorSimilarityFunction;
+import org.apache.lucene.queries.function.FunctionScoreQuery;
+import org.apache.lucene.search.DoubleValuesSource;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.TopDocs;
+import org.elasticsearch.index.mapper.vectors.VectorSimilarityByteValueSource;
+import org.elasticsearch.index.mapper.vectors.VectorSimilarityFloatValueSource;
+import org.elasticsearch.search.profile.query.QueryProfiler;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * Wraps an internal query to rescore the results using a similarity function over the original, non-quantized vectors of a vector field
+ */
+public class RescoreKnnVectorQuery extends Query implements ProfilingQuery {
+    private final String fieldName;
+    private final byte[] byteTarget;
+    private final float[] floatTarget;
+    private final VectorSimilarityFunction vectorSimilarityFunction;
+    private final Integer k;
+    private final Query innerQuery;
+
+    private long vectorOpsCount;
+
+    public RescoreKnnVectorQuery(
+        String fieldName,
+        byte[] byteTarget,
+        VectorSimilarityFunction vectorSimilarityFunction,
+        Integer k,
+        Query innerQuery
+    ) {
+        this.fieldName = fieldName;
+        this.byteTarget = byteTarget;
+        this.floatTarget = null;
+        this.vectorSimilarityFunction = vectorSimilarityFunction;
+        this.k = k;
+        this.innerQuery = innerQuery;
+    }
+
+    public RescoreKnnVectorQuery(
+        String fieldName,
+        float[] floatTarget,
+        VectorSimilarityFunction vectorSimilarityFunction,
+        Integer k,
+        Query innerQuery
+    ) {
+        this.fieldName = fieldName;
+        this.byteTarget = null;
+        this.floatTarget = floatTarget;
+        this.vectorSimilarityFunction = vectorSimilarityFunction;
+        this.k = k;
+        this.innerQuery = innerQuery;
+    }
+
+    @Override
+    public Query rewrite(IndexSearcher searcher) throws IOException {
+        assert byteTarget == null ^ floatTarget == null : "Either byteTarget or floatTarget must be set";
+
+        final DoubleValuesSource valueSource;
+        if (byteTarget != null) {
+            valueSource = new VectorSimilarityByteValueSource(fieldName, byteTarget, vectorSimilarityFunction);
+        } else {
+            valueSource = new VectorSimilarityFloatValueSource(fieldName, floatTarget, vectorSimilarityFunction);
+        }
+        FunctionScoreQuery functionScoreQuery = new FunctionScoreQuery(innerQuery, valueSource);
+        Query query = searcher.rewrite(functionScoreQuery);
+
+        if (k == null) {
+            // No need to calculate top k - let the request size limit the results
+            return query;
+        }
```

Comment:
Here, we now miss the rescoring of ALL the vectors returned internally. We need to have those.